File: src/vscripts/traps/fire_trap_trigger.ts
export {};

declare const _G: Record<string, any>;
declare const thisEntity: CBaseEntity;

let triggerActive = true;
let n = 0;

function OnStartTouch(this: void, trigger: { activator: CDOTA_BaseNPC }): void {
    const triggerName = thisEntity.GetName();
    const model = triggerName + "_model";
    const npc = Entities.FindByNameWithin(null as any, triggerName + "_npc", thisEntity.GetAbsOrigin(), 4000) as CDOTA_BaseNPC | undefined;
    if (!npc) {
        return;
    }
    if ((npc as any).fNextAttackTime && (npc as any).fNextAttackTime >= GameRules.GetGameTime()) {
        return;
    }
    const target = Entities.FindByNameWithin(null as any, triggerName + "_target", thisEntity.GetAbsOrigin(), 4000) as CDOTA_BaseNPC | undefined;
    const fireTrap = npc.FindAbilityByName("breathe_fire");
    const fire_delay = 0.03;
    const reset_time = 1;
    const activator = trigger.activator;

    npc.SetContextThink(
        "fire_trap_anim",
        () => {
            DoEntFire(model, "SetAnimation", "bark_attack", 0.4, undefined as any, undefined as any);
            npc.SetContextThink(
                "fire_trap_cast",
                () => {
                    (fireTrap as any).activator = activator;
                    npc.CastAbilityOnPosition(target!.GetOrigin(), fireTrap!, -1);
                    return undefined;
                },
                0.3
            );
            return undefined;
        },
        fire_delay
    );

    const heroIndex = trigger.activator.GetEntityIndex();
    const heroHandle = EntIndexToHScript(heroIndex);
    print("Trap Button Trigger Entered", (heroHandle as CDOTA_BaseNPC).GetUnitName());

    (npc as any).KillerToCredit = heroHandle;
    triggerActive = false;
    (npc as any).fRefireTime = 1.8;
    (npc as any).fQuickRefireTime = reset_time;
    (npc as any).bNextAttackIsNormal = false;
    (npc as any).nQuickRefires = 0;
    (npc as any).fNextAttackTime = GameRules.GetGameTime() + (npc as any).fQuickRefireTime;
}

function OnEndTouch(this: void, trigger: any): void {
    // no-op
}

_G.OnStartTouch = OnStartTouch;
_G.OnEndTouch = OnEndTouch;

File: game/A1/scripts/vscripts/traps/fire_trap_trigger.lua

--[[ fire_trap_trigger.lua ]]

--按钮式火焰陷阱

--踩陷阱后会延迟x秒发射火焰

local triggerActive = true
local n=0
function OnStartTouch(trigger)
    local triggerName = thisEntity:GetName()
    local model = triggerName .. "_model"
    -- local npc = Entities:FindByName( nil, triggerName .. "_npc" )
    local npc = Entities:FindByNameWithin(nil, triggerName .. "_npc", thisEntity:GetAbsOrigin(), 4000)
    if not npc then
        return
    end
    if npc.fNextAttackTime and npc.fNextAttackTime>=GameRules:GetGameTime() then
        return
    end
    -- local target = Entities:FindByName( nil, triggerName .. "_target" )
    local target = Entities:FindByNameWithin(nil, triggerName .. "_target", thisEntity:GetAbsOrigin(), 4000)
    local fireTrap = npc:FindAbilityByName("breathe_fire")

    local fire_delay = 0.03
    local reset_time = 1
    
    local activator = trigger.activator

    
    npc:GameTimer(fire_delay, function()
        DoEntFire( model, "SetAnimation", "bark_attack", .4, self, self )
        npc:GameTimer(0.3, function()
            fireTrap.activator = activator
            npc:CastAbilityOnPosition(target:GetOrigin(), fireTrap, -1 )
        end)
    end)
    
    -- npc:CastAbilityOnPosition(target:GetOrigin(), fireTrap, -1 )
    -- npc:SetContextThink( "ResetButtonModel", function() ResetButtonModel() end, 0 )
    
    local heroIndex = trigger.activator:GetEntityIndex()
    local heroHandle = EntIndexToHScript(heroIndex)


    print("Trap Button Trigger Entered",heroHandle:GetUnitName())

    npc.KillerToCredit = heroHandle
    triggerActive = false
    npc.fRefireTime = 1.8
    npc.fQuickRefireTime = reset_time
    npc.bNextAttackIsNormal = false

    npc.nQuickRefires = 0
    npc.fNextAttackTime = GameRules:GetGameTime() + npc.fQuickRefireTime
end

function OnEndTouch(trigger)
   
end

